L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,501 @@
+  // know.
+  //
+  // request: the verification information associated with this request
+  // callback: This will only take effect if the verifier is asynchronous.
+  //           The function that gRPC will invoke when the verifier has already
+  //           completed its asynchronous check. Callers can use this function
+  //           to perform any additional checks. The input parameter of the
+  //           std::function indicates the status of the verifier check.
+  // sync_status: This will only be useful if the verifier is synchronous.
+  //              The status of the verifier as it has already done it's
+  //              synchronous check.
+  // return: return true if executed synchronously, otherwise return false
+  bool Verify(TlsCustomVerificationCheckRequest* request,
+              std::function<void(Status)> callback,
+              Status* sync_status);
+
+  // Cancels a verification request previously started via Verify().
+  // Used when the connection attempt times out or is cancelled while an async
+  // verification request is pending.
+  //
+  // request: the verification information associated with this request
+  void Cancel(TlsCustomVerificationCheckRequest* request);
+};
+
+// A CertificateVerifier that doesn't perform any additional checks other than
+// certificate verification, if specified.
+// Note: using this solely without any other authentication mechanisms on the
+// peer identity will leave your applications to the MITM(Man-In-The-Middle)
+// attacks. Users should avoid doing so in production environments.
+class NoOpCertificateVerifier : public CertificateVerifier {
+ public:
+  NoOpCertificateVerifier();
+};
+
+// A CertificateVerifier that will perform hostname verification, to see if the
+// target name set from the client side matches the identity information
+// specified on the server's certificate.
+class HostNameCertificateVerifier : public CertificateVerifier {
+ public:
+  HostNameCertificateVerifier();
+};
+```
+
+#### Custom Chain Building
+```c++
+class CustomChainBuilderInterface {
+public:
+ virtual ~CustomChainBuilderInterface() = default;
+
+ // Returns the verified DER-encoded certificate chain, ordered from leaf to
+ // root. Both the leaf and the root are included. Returns an error status if
+ // verification fails for any reason.
+ virtual absl::StatusOr<std::vector<absl::string_view>> BuildAndVerifyChain(const std::vector<absl::string_view>& peer_cert_chain_der) = 0;
+}
+```
+
+The default behavior for chain building is to defer to the underlying SSL library, which has a built-in chain building API that has been hardened over many years of use with the web PKI. For example, [X509_verify_cert](https://www.openssl.org/docs/man1.1.1/man3/X509_verify_cert.html) is the implementation in OpenSSL that builds a chain and verifies that it is trusted by one of the root certificates.
+
+```c++
+class SPIFFEFederationChainBuilder : public CustomChainBuilderInterface {
```

Comment:
Similar to the earlier examples, should we just have the constructor only and omit the other methods?

Reply:
I don't think this was addressed: we only have the destructor, rather than only having the constructor.

Reply:
Good catch, fixed